subsys/bluetooth/host/classic/a2dp.c

```diff
@@ -285,166 +321,104 @@ static void bt_a2dp_media_data_callback(
+	struct bt_avdtp_sep *sep, uint8_t *errcode,
+	int (*req_fun)(struct bt_a2dp_stream *stream, uint8_t *rsp_err_code),
+	void (*done_fun)(struct bt_a2dp_stream *stream),
+	bool clear_stream)
 {
 	struct bt_a2dp_ep *ep;
 	struct bt_a2dp_stream *stream;
-	struct bt_a2dp_stream_ops *ops;
 
 	*errcode = 0;
 	ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
 	if (ep->stream == NULL) {
 		*errcode = BT_AVDTP_BAD_ACP_SEID;
-		return -1;
+		return -EINVAL;
 	}
 	stream = ep->stream;
 
-	if ((a2dp_cb != NULL) && (a2dp_cb->establish_req != NULL)) {
+	if (req_fun != NULL) {
 		uint8_t rsp_err_code;
 		int err;
 
-		err = a2dp_cb->establish_req(stream, &rsp_err_code);
+		err = req_fun(stream, &rsp_err_code);
 		if (err) {
 			*errcode = rsp_err_code;
 		}
 	}
 
-	ops = stream->ops;
 	if (*errcode == 0) {
-		if ((ops != NULL) && (stream->ops->established != NULL)) {
-			stream->ops->established(stream);
-		}
-	}
-
-	return (*errcode == 0) ? 0 : -1;
-}
-
-static int a2dp_start_ind(struct bt_avdtp *session, struct bt_avdtp_sep *sep, uint8_t *errcode)
-{
-	struct bt_a2dp_ep *ep;
-	struct bt_a2dp_stream *stream;
-	struct bt_a2dp_stream_ops *ops;
-
-	*errcode = 0;
-	ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
-	if (ep->stream == NULL) {
-		*errcode = BT_AVDTP_BAD_ACP_SEID;
-		return -1;
-	}
-	stream = ep->stream;
-
-	if ((a2dp_cb != NULL) && (a2dp_cb->start_req != NULL)) {
-		uint8_t rsp_err_code;
-		int err;
-
-		err = a2dp_cb->start_req(stream, &rsp_err_code);
-		if (err) {
-			*errcode = rsp_err_code;
+		if (clear_stream) {
+			ep->stream = NULL;
 		}
-	}
 
-	ops = stream->ops;
-	if (*errcode == 0) {
-		if ((ops != NULL) && (stream->ops->started != NULL)) {
-			stream->ops->started(stream);
+		if (done_fun != NULL) {
+			done_fun(stream);
 		}
 	}
 
-	return (*errcode == 0) ? 0 : -1;
+	return (*errcode == 0) ? 0 : -EINVAL;
 }
 
-static int a2dp_close_ind(struct bt_avdtp *session, struct bt_avdtp_sep *sep, uint8_t *errcode)
+static int a2dp_open_ind(struct bt_avdtp *session, struct bt_avdtp_sep *sep, uint8_t *errcode)
 {
-	struct bt_a2dp_ep *ep;
-	struct bt_a2dp_stream *stream;
-	struct bt_a2dp_stream_ops *ops;
-
-	*errcode = 0;
-	ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
-	if (ep->stream == NULL) {
-		*errcode = BT_AVDTP_BAD_ACP_SEID;
-		return -1;
-	}
-	stream = ep->stream;
-
-	if ((a2dp_cb != NULL) && (a2dp_cb->release_req != NULL)) {
-		uint8_t rsp_err_code;
-		int err;
-
-		err = a2dp_cb->release_req(stream, &rsp_err_code);
-		if (err) {
-			*errcode = rsp_err_code;
-		}
-	}
-
-	ops = stream->ops;
-	if (*errcode == 0) {
-		if ((ops != NULL) && (stream->ops->released != NULL)) {
-			stream->ops->released(stream);
-		}
-	}
+	struct bt_a2dp_ep *ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
+
+	BT_ASSERT_MSG(sep, "Invalid sep");
+	return a2dp_ctrl_ind(session, sep, errcode,
+		a2dp_cb != NULL ? a2dp_cb->establish_req : NULL,
+		(ep->stream != NULL && ep->stream->ops != NULL) ?
+		ep->stream->ops->established : NULL,
+		false);
```

Comment:
Sorry, but this is just unreadable. Could you introduce some helper variables which you prepare before the call to `a2dp_ctrl_ind()`, so that the actual call itself doesn't contain any conditionals. The resulting compiled code will most like be exactly the same since the compiler will optimize this either way. Also, please make sure you're using correct alignment of N tabs + 0-7 spaces, so that the split lines start right after the ( on the first line.

Reply:
Updated all the cases.